downstairs/src/region.rs

```diff
@@ -1018,34 +1096,62 @@ impl Extent {
          * a checksum.
          */
 
-        let mut writes_to_skip: Vec<u64> = Vec::new();
+        // If `only_write_written`, we need to skip writing to blocks that
+        // already contain data. We'll first query the metadata to see which
+        // blocks have hashes
+        let mut writes_to_skip = HashSet::new();
         if only_write_unwritten {
-            for write in writes {
-                if !inner.get_hashes(write.offset.value).unwrap().is_empty() {
-                    writes_to_skip.push(write.offset.value);
+            let mut write_run_start = 0;
+            while write_run_start < writes.len() {
+                let first_write = writes[write_run_start];
+
+                // Starting from the first write in the potential run, we scan
+                // forward until we find a write with a block that isn't
+                // contiguous with the request before it. Since we're counting
+                // pairs, and the number of pairs is one less than the number of
+                // writes, we need to add 1 to get our actual run length.
+                let n_contiguous_writes = writes[write_run_start..]
+                    .windows(2)
+                    .take_while(|wr_pair| {
+                        wr_pair[0].offset.value + 1 == wr_pair[1].offset.value
+                    })
+                    .count()
+                    + 1;
+
+                // Query hashes for the write range.
+                // XXX we should add a query that doesnt actually give us back
```

Comment:
Is this XXX something for the future, or for this PR? :)

Reply:
Oh I haven't heard of that, no. I like that though, I think I'll try following that convention.